File: src/main/java/com/babylonml/backend/training/optimizer/GradientOptimizer.java
package com.babylonml.backend.training.optimizer;

import com.babylonml.backend.training.execution.TrainingExecutionContext;
import com.babylonml.backend.training.operations.Operation;
import it.unimi.dsi.fastutil.ints.IntImmutableList;
import org.jspecify.annotations.NonNull;

import java.util.List;

public interface GradientOptimizer {
    void optimize(TrainingExecutionContext executionContext,
                  float[] matrix, int matrixOffset, IntImmutableList shape, float[] gradient,
                  int gradientOffset, float learningRate, Operation operation);

     @NonNull
     List<IntImmutableList> calculateRequiredMemoryAllocations(IntImmutableList shape);
}

File: src/main/kotlin/com/babylonml/backend/training/optimizer/GradientOptimizer.kt
package com.babylonml.backend.training.optimizer

import com.babylonml.backend.training.execution.TrainingExecutionContext
import com.babylonml.backend.training.operations.Operation
import it.unimi.dsi.fastutil.ints.IntImmutableList

interface GradientOptimizer {
    fun optimize(
        executionContext: TrainingExecutionContext,
        matrix: FloatArray, matrixOffset: Int, shape: IntImmutableList, gradient: FloatArray,
        gradientOffset: Int, learningRate: Float, operation: Operation
    )

    fun calculateRequiredMemoryAllocations(shape: IntImmutableList): List<IntImmutableList>
}
